passive/detect_csp_notif_and_reportonly.js

```diff
@@ -0,0 +1,68 @@
+
+Links:
+- http://content-security-policy.com
+- http://www.html5rocks.com/en/tutorials/security/content-security-policy
+- http://www.html5rocks.com/en/tutorials/security/content-security-policy/#reporting
+
+Author: dominique.righetto@gmail.com
+*/
+
+function scan(ps, msg, src) {
+	//Docs on alert raising function:
+	// raiseAlert(risk, int confidence, String name, String description, String uri, 
+	//		String param, String attack, String otherInfo, String solution, String evidence, 
+	//		int cweId, int wascId, HttpMessage msg)
+	// risk: 0: info, 1: low, 2: medium, 3: high
+	// confidence: 0: falsePositive, 1: low, 2: medium, 3: high, 4: confirmed
+
+	//Common variables
+	var cweId = 200; 
+	var wascId = 13;
+	var url = msg.getRequestHeader().getURI().toString();	
+	var cspHeaderNames = ["Content-Security-Policy", "X-Content-Security-Policy", "X-Webkit-CSP", "Content-Security-Policy-Report-Only"];
+	var cspReportInstruction = "report-uri";
+
+	//Response headers collection
+	var responseHeaders = msg.getResponseHeader();
+
+	//Detect and analyze presence of the CSP headers
+	for(var i = 0 ; i < cspHeaderNames.length ; i++){
+		var headerName = cspHeaderNames[i];
+		if(responseHeaders.getHeaders(headerName)){
+			//Check if the header values (policies) contains the CSP reporting instruction
+			var headerValues = responseHeaders.getHeaders(headerName).toArray();
+			for(var j = 0 ; j < headerValues.length ; j++){
+				var cspPolicies = headerValues[j].toLowerCase();
+				if(cspPolicies.indexOf(cspReportInstruction) != -1){
+					//Extract the URL to which any CSP violations are reported
+					//In CSP specification, policies are separated by ';'
+					var startPosition = cspPolicies.search(cspReportInstruction);
+					var tmp = cspPolicies.substring(startPosition);
+					var endPosition = tmp.indexOf(";");
+					if(endPosition != -1){
+						var reportUrl = tmp.substring(0, endPosition);	
+					}else{
+						var reportUrl = tmp;
+					}
+					reportUrl = reportUrl.replace(cspReportInstruction,"").trim();					
+					//Raise info alert
+					var cspWorkingMode = (!headerName.toLowerCase().endsWith("-report-only")) ? "BLOCKING" : "REPORTING";
+					var description = "The current site CSP policies defined by HTTP response header '" + headerName + "' (behaving in " + cspWorkingMode + " mode) report violation to '" + reportUrl + "'.";
+					var infoLinkRef = "https://developer.mozilla.org/en-US/docs/Web/Security/CSP/Using_CSP_violation_reports";	
+					var solution = "Site owner will be notified at each policies violations, so, start by analyzing if a real monitoring of the notifications is in place before to use fuzzing or to be more aggressive.";				
+					ps.raiseAlert(0, 4, "Content Security Policy violations reporting enabled", description, 
+						url, "HTTP response header '" + headerName + "'", "Non applicable", infoLinkRef, solution, "CSP policies: " + headerValues[j], cweId, wascId, msg);					
```

Comment:
- For passive scanners the Attack field should be left empty;
 - Evidence could just be `headerValues[j]`, it would be automatically highlighted in response tab;
 - The tabs at the end of the line could be removed.